src/text/mg_passes.1
.\"------------------------------------------------------------
.\" Id - set Rv,revision, and Dt, Date using rcs-Id tag.
.de Id
.ds Rv \\$3
.ds Dt \\$4
..
.Id $Id: mg_passes.1,v 1.1 1994/09/19 06:03:05 tes Exp $
.\"------------------------------------------------------------
.TH mg_passes 1 \*(Dt CITRI
.SH NAME
mg_passes \- builds mg databases
.SH SYNOPSIS
.B mg_passes
[
.B \-h
]
[
.B \-G
]
[
.B \-S
]
[
.B \-D
]
[
.B \-W
]
.if n .ti +9n
[
.BR \-1 " |"
.BR \-2 " |"
.B  \-3
]
[
.BI \-C " compstatpoint"
]
.if n .ti +9n
[
.BI \-n " tracename"
]
.if t .ti +.5i
[
.BI \-b " bufsize"
]
[
.BI \-m " memlimit"
]
.if n .ti +9n
[
.BI \-c " numchunks"
]
[
.BI \-s " stemmethod"
]
[
.BI \-t " tracepos"
]
.if n .ti +9n
[
.B \-T1
]
[
.B \-T2
]
.if t .ti +.5i
[
.B \-I1
]
[
.B \-I2
]
.if n .ti +9n
[
.BI \-d " directory"
]
.BI \-f " name"
[
.I filename(s)
]
.SH DESCRIPTION
.B mg_passes
is the program that does most of the work when building
.BR mg (1)
database systems.  The input documents can come from either
.I stdin
or from a list of files on the command line.  Individual documents
must be separated with control-B characters.  In general,
.B mg_passes
must be run twice to build a database, first with the
.B \-T1
and
.B \-I1
options, and second with the
.B \-T2
and
.B \-I2
options.  Several other programs must be run in order to get an
.BR mg (1)
database that is ready for the
.BR mgquery (1)
program.  The
.SB EXAMPLE
section below gives an example of how to build a complete
.BR mg (1)
database.
.SH OPTIONS
Options may appear in any order, but the
.IR filename(s) ,
if specified, must be last.
.TP "\w'\fB\-C\fP \fIcompstatpoint\fP'u+2n"
.B \-h
This displays a usage line on
.IR stderr .
.TP
.B \-G
Treat SGML tags as non-words when building the inverted file.  An SGML
tag is anything between angle brackets, i.e., `<' and `>'.
.TP
.B \-S
This option causes a special pass to be executed.  It is up to a user
to modify
.I mg.special.c
in the source code to do something with the documents it is given.
.TP
.B \-D
If
.B mg_passes
fails, then print the document that caused the failure to the trace
file if tracing is active, or to
.I stderr
if it is not.
.TP
.B \-W
This option enables the generation of the weights file when
.B \-I2
is specified.  It causes
.B \-I2
to use a little more memory and CPU.
.TP
.B \-1
Produce a level-1 inverted file.  This option is only useful when
specified with
.BR "\-I1 ".
A level-1 inverted file makes it possible for
.BR mgquery (1)
to do Boolean queries.  Ranked queries can still be done,
although the quality of the ranking is abysmal.
.TP
.B \-2
Produce a level-2 inverted file.  This option is only useful when
specified with
.BR "\-I1 ".
This is the default when neither
.BR \-1 ", " "\-2 " "or " \-3
is specified.
A level-2 inverted file makes it possible for
.BR mgquery (1)
to do Boolean queries and cosine-ranked queries.
.TP
.B \-3
Produce a level-3 inverted file.  This option is only useful when
specified with
.BR "\-I1 ".
This has been implemented to enable paragraph-level inversion.
Paragraphs are delimited by control-C characters in the source text.
.TP
.BI \-C " compstatpoint"
This option causes statistics on the compression performance to be
output to a file called
.IR *.compression.stats .
.I compstatpoint
specifies the interval between outputting each line of statistics.  The
units of
.I compstatpoint
are kilobytes of source text.  E.g., if
.I compstatpoint
is 10, then a line is output to the file every 10 KB of input
source.  Each line of the file consists of 4 numbers The first number
is the amount of input text, in bytes, processed so far.  The second
number is the amount of input text, in bytes, processed since the
last line was output to the file.  The third number is the number of
output bytes generated since the last line was output to the file, and
the fourth number is the compression achieved since the last line was
output, i.e., the third number divided by the second number.
.TP
.BI \-n " tracename"
This specifies the filename to use for the trace log, if tracing is
enabled using the
.B \-t
option.  If
.BI \-n " tracename"
is not given and tracing is enabled, a default trace filename will be
used.
.TP
.BI \-s " stemmethod"
This specifies the method to use to \*(lqstem\*(rq the words in the
inverted file dictionary.  This is a bit mask specifying the
operations to do on words as they are parsed out of the text, where
bit number 0 is the low-order (rightmost) bit.  Bit 0 does case
folding, and bit 1 does simple stemming, so the value 3 for
.I stemmethod
does both case folding and stemming.
.TP
.BI \-b " bufsize"
Specify the size of the document buffer in kilobytes.  If any document
is larger than
.IR bufsize ,
the program will abort with an error message.  This should probably be
replaced with some system which automatically increases the buffer
size as required.  The default size is 3072 KB (3 MB).
.TP
.BI \-m " memlimit"
Maximum amount of memory to use for the pass-2 file inversion in
megabytes.  This option is only useful when used in conjunction with
the option
.BR \-I1 .
The larger this value, the faster the pass-2 inversion will proceed.
The default value is 5 MB.
.TP
.BI \-c " numchunks"
The maximum number of inversion chunks to write to disk.  Each chunk
will be approximately as large as
.IR memlimit .
This option is only useful when used in conjunction with the option
.BR \-I2 .
The larger this value, the faster the pass-2 inversion will proceed.
The default value is 5 MB.
.TP
.BI \-t " tracepos"
This option activates tracing.  A line will be generated in the
trace file for every
.I tracepos
input bytes processed.  The default name for the trace file can be
overridden using the
.BI \-n " tracename"
option.
.TP
.B \-T1
Generate the
.I *.text.stats
file.
.TP
.B \-T2
Generate the
.IR *.text ,
.IR *.text.idx ,
and possibly the
.I *.text.dict.aux
files.  Using this option requires that the
.I *.text.dict
file be present.
.TP
.B \-I1
Generate the
.IR *.invf.dict ,
.IR *.invf.chunk ,
and
.I *.invf.chunk.trans
files.
.TP
.B \-I2
Generate the
.I *.invf
and
.I *.invf.idx
files.  Using this option requires
that the
.IR *.invf.dict.hash ,
.IR *.invf.chunk ,
and
.I *.invf.chunk.trans
files
be present.  The
.I *.invf.dict.hash
file is generated by
.BR mg_perf_hash_build (1)
from the
.I *.invf.dict.build
file.  If the
.B \-W
option is specified, the
.I *.weight
file will also be generated.
.TP
.BI \-d " directory"
This specifies the directory where the document collection is to be
written.
.TP
.BI \-f " name"
This specifies the base name of the document collection that will be
created.
.TP
.I filename(s)
This specifies the source text. If this is not specified, then the
program expects the source text from
.IR stdin .
.SH EXAMPLE
What follows is a UNIX
.BR csh (1)
script as an example of how to build an
.BR mg (1)
document collection.
.LP
.nf
.DT
.ft B
.I #! /bin/csh
.I
# The first argument on the command line specifies the
.I
#   source of the text
set source = ($1)
.PP
.I
# The second argument is the name of the collection
set text = ($2)
.PP
.I
# Create *.text.stats,  *.invf.dict.build,
.I
#   *.invf.chunk and *.invf.chunks.trans
${source} | mg_passes -T1 -I1 -m 1 -t 1 -f ${text}
.PP
.I
# Create *.text.dict
mg_compression_dict -f ${text}
.PP
.I
# Create *.invf.dict.hash
mg_perf_hash_build -f ${text}
.PP
.I
# Create *.text,  *.text.idx,
.I
#   *.invf and *.invf.idx
${source} | mg_passes -T2 -I2 -c 2 -t 1 -f ${text}
.PP
.I
# Create *.text.idx.wgt and *.weight.approx
mg_weights_build -f ${text} -b 8
.PP
.I
# Create *.invf.dict
mg_invf_dict -f ${text} -b 4096
.PP
.I
# Create *.text.dict
mg_fast_comp_dict -f ${text}
.ft R
.fi
.SH ENVIRONMENT
.TP "\w'\fBMGDATA\fP'u+2n"
.SB MGDATA
If this environment variable exists, then its value is used as the
default directory where the
.BR mg (1)
collection files are.  If this variable does not exist, then the
directory \*(lq\fB.\fP\*(rq is used by default.  The command line
option
.BI \-d " directory"
overrides the directory in
.BR MGDATA .
.SH FILES
.TP 20
.B *.invf
Inverted file.
.TP
.B *.invf.chunk
Inverted file chunk descriptor file.  When the inverted file is
created it is created in chunks that use no more than a set amount of
memory.  This file describes those chunks.
.TP
.B *.invf.chunk.trans
Word-occurrence-order to lexical-order translation file.  The
.B *.invf.chunk
file is written in word-occurrence order but is required by
.B \-I2
to be in lexical order.
.TP
.B *.invf.dict.build
Compressed stemmed dictionary.
.TP
.B *.invf.dict.hash
Data for an order-preserving perfect hash function.
.TP
.B *.invf.idx
The index into the inverted file.
.TP
.B *.weight
The exact weights file.
.TP
.B *.text
Compressed documents.
.TP
.B *.text.stats
Statistics about the text.
.TP
.B *.text.dict
Compressed compression dictionary.
.TP
.B *.text.idx
Index into the compressed documents.
.TP
.B *.trace
The default trace file.
.TP
.B *.compression.stats
Statistics about the compression of the text.
.SH "SEE ALSO"
.na
.BR mg (1),
.BR mg_compression_dict (1),
.BR mg_fast_comp_dict (1),
.BR mg_get (1),
.BR mg_invf_dict (1),
.BR mg_invf_dump (1),
.BR mg_invf_rebuild (1),
.BR mg_perf_hash_build (1),
.BR mg_text_estimate (1),
.BR mg_weights_build (1),
.BR mgbilevel (1),
.BR mgbuild (1),
.BR mgdictlist (1),
.BR mgfelics (1),
.BR mgquery (1),
.BR mgstat (1),
.BR mgtic (1),
.BR mgticbuild (1),
.BR mgticdump (1),
.BR mgticprune (1),
.BR mgticstat (1).


src/text/mg_passes.c
/**************************************************************************
 *
 * mg_passes.c -- Driver for the various passes
 * Copyright (C) 1994  Neil Sharman
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id: mg_passes.c,v 1.3 1994/10/20 03:56:57 tes Exp $
 *
 **************************************************************************/

#include "sysfuncs.h"

#ifdef HAVE_MALLINFO
# include <malloc.h> 
#endif

#include "memlib.h"
#include "messages.h"
#include "timing.h"

#include "longlong.h"

#include "mg_files.h"
#include "mg.h"
#include "build.h"
#include "text.h"
#include "stemmer.h"


/*
   $Log: mg_passes.c,v $
   * Revision 1.3  1994/10/20  03:56:57  tes
   * I have rewritten the boolean query optimiser and abstracted out the
   * components of the boolean query.
   *
   * Revision 1.2  1994/09/20  04:41:52  tes
   * For version 1.1
   *
 */

static char *RCSID = "$Id: mg_passes.c,v 1.3 1994/10/20 03:56:57 tes Exp $";

#define MAX_PASSES 5

#define SPECIAL 1
#define TEXT_PASS_1 2
#define TEXT_PASS_2 4
#define IVF_PASS_1 8
#define IVF_PASS_2 16

#define MIN_BUF 8192
#define	TERMRECORD	'\002'

unsigned long buf_size = 3 * 1024 * 1024;	/* 3Mb */
unsigned long invf_buffer_size = 5 * 1024 * 1024;	/* 5Mb */
unsigned long ChunkLimit = 0;
char InvfLevel = 2;
char SkipSGML = 0;
char MakeWeights = 0;
FILE *Comp_Stats = NULL;
int comp_stat_point = 0;
mg_ullong bytes_processed = 0;
mg_ullong bytes_received = 0;
int stem_method = 0;

static char Passes = 0;
static unsigned long trace = 0;
static int Dump = 0;
static char **files = NULL;
static int num_files = 0;
static char *trace_name = NULL;


typedef struct pass_data
  {
    char *name;
    int (*init) (char *);
    int (*process) (u_char *, int);
    int (*done) (char *);
#ifdef HAVE_TIMES
    clock_t init_time;
    clock_t process_time;
    clock_t done_time;
#else
    struct timeval init_time;
    struct timeval process_time;
    struct timeval done_time;
#endif
  }
pass_data;

#ifdef HAVE_TIMES
#define NULL_TIMES 0, 0, 0
#else
#define NULL_TIMES {0, 0}, {0, 0}, {0, 0}
#endif

static pass_data PassData[MAX_PASSES] =
{
  {"special", init_special, process_special, done_special, NULL_TIMES},
  {"text.pass1", init_text_1, process_text_1, done_text_1, NULL_TIMES},
  {"text.pass2", init_text_2, process_text_2, done_text_2, NULL_TIMES},
  {"ivf.pass1", init_ivf_1, process_ivf_1, done_ivf_1, NULL_TIMES},
  {"ivf.pass2", init_ivf_2, process_ivf_2, done_ivf_2, NULL_TIMES},
};

static char *usage_str = "\nUSAGE:\n"
"  %s [-h] [-G] [-D] [-1|-2|-3] [-T1] [-T2] [-I1] [-I2] [-N1]\n"
"  %*s [-N2] [-W] [-S] [-b buffer-size] [-d dictionary-directory]\n"
"  %*s [-t trace-point Mb] [-m invf-memory] [-c chunk-limit]\n"
"  %*s [-n trace-name] [-C comp-stat-size] [-s stem_method] -f doc-collection-name\n";









static void 
usage (char *err)
{
  if (err)
    Message (err);
  fprintf (stderr, usage_str, msg_prefix, strlen (msg_prefix), "",
	   strlen (msg_prefix), "", strlen (msg_prefix), "");
  exit (1);
}




#if 0
static char *
str_comma (unsigned long u)
{
  static char buf[20];
  unsigned long a, b, c, d;
  a = u / 1000000000;
  u -= a * 1000000000;
  b = u / 1000000;
  u -= b * 1000000;
  c = u / 1000;
  u -= c * 1000;
  d = u;

  if (a)
    sprintf (buf, "%u,%03u,%03u,%03u", a, b, c, d);
  else if (b)
    sprintf (buf, "%u,%03u,%03u", b, c, d);
  else if (c)
    sprintf (buf, "%u,%03u", c, d);
  else
    sprintf (buf, "%u", d);
  return (buf);
}
#endif




int 
open_next_file (int in_fd)
{
  if (in_fd > 0)
    close (in_fd);
  if (num_files == 0)
    return (-1);
  if ((in_fd = open (files[0], O_RDONLY)) == -1)
    FatalError (1, "Cannot open %s", files[0]);
  files++;
  num_files--;
  return (in_fd);
}


static void 
driver (int in_fd, FILE * Trace, char *file_name)
{
  int pass, num = 1;

  char *buffer = Xmalloc (buf_size);
  unsigned long num_docs = 0;
  unsigned long block_bytes = 0;
  register int buf_left = buf_size;
  register char *look_pos = buffer;
  register char *end_pos = buffer;

  ProgTime StartTime, InitTime, ProcTime, DoneTime;

  GetTime (&StartTime);

  for (pass = 0; pass < MAX_PASSES; pass++)
    if (Passes & (1 << pass))
      {
	pass_data *pd = &PassData[pass];
#ifdef HAVE_TIMES
	struct tms tims;
	times (&tims);
	pd->init_time -= tims.tms_utime + tims.tms_stime;
#else
	struct rusage ru;

	getrusage (RUSAGE_SELF, &ru);
	pd->init_time.tv_sec -= ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	pd->init_time.tv_usec -= ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
#endif
	if (pd->init (file_name) == COMPERROR)
	  FatalError (1, "Error during init of \"%s\"", pd->name);

#ifdef HAVE_TIMES
	times (&tims);
	pd->init_time += tims.tms_utime + tims.tms_stime;
#else
	getrusage (RUSAGE_SELF, &ru);
	pd->init_time.tv_sec += ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	pd->init_time.tv_usec += ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
	time_normalise (&pd->init_time);
#endif
      }

  GetTime (&InitTime);
  for (;;)
    {
      int len = 0;
      char *base = look_pos;

      while (look_pos != end_pos && *look_pos != TERMRECORD)
	look_pos++;

      while (look_pos == end_pos)
	{
	  if (buf_left < MIN_BUF)
	    {
	      bcopy (base, buffer, end_pos - base);
	      look_pos = buffer + (end_pos - base);
	      buf_left = buf_size - (end_pos - base);
	      end_pos = look_pos;
	      base = buffer;
	    }
	  if (buf_left)
	    {
	      num = read (in_fd, end_pos, buf_left);
	      if (num == 0)
		if ((in_fd = open_next_file (in_fd)) != -1)
		  num = read (in_fd, end_pos, buf_left);
	      bytes_received += num;
	      buf_left -= num;
	      end_pos += num;
	    }
	  while (look_pos != end_pos && *look_pos != TERMRECORD)
	    look_pos++;
	  if (buf_left == 0 && base == buffer && look_pos == end_pos)
	    {
	      Message ("Unable to find document terminator (i.e ^B)"
		       " in the document");
	      FatalError (1, "The document is in excess of %d chars long",
			  look_pos - base);
	    }
	  if (!num)
	    break;
	}
      len = look_pos++ - base;

      if (!num && base == end_pos)
	break;

      bytes_processed += len;

      if (!len)
	Message ("Warning : Processing zero length document");

      for (pass = 0; pass < MAX_PASSES; pass++)
	if (Passes & (1 << pass))
	  {
	    register pass_data *pd = &PassData[pass];

#ifdef HAVE_TIMES
	    struct tms tims;
	    times (&tims);
	    pd->process_time -= tims.tms_utime + tims.tms_stime;
#else
	    struct rusage ru;
	    register struct timeval *tv = &pd->process_time;

	    getrusage (RUSAGE_SELF, &ru);
	    tv->tv_sec -= ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	    tv->tv_usec -= ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
#endif

	    if (pd->process ((u_char *) base, len) == COMPERROR)
	      {
		Message ("Error during processing of \"%s\"", pd->name);
		if (Dump || Trace)
		  {
		    int i;
		    FILE *f = Trace ? Trace : stderr;
		    fprintf (f, "-=- * -=- * -=- * -=- * -=- * -=- * -=-\n");
		    for (i = 0; i < len; i++)
		      {
			char ch = base[i];
			if (ch == '\1' || ch == '\2')
			  ch = '\n';
			putc (ch, f);
		      }
		    fprintf (f, "-=- * -=- * -=- * -=- * -=- * -=- * -=-\n");
		  }
		if (Trace)
		  fprintf (Trace, "%11" ULL_FS " bytes |%7lu docs | %s\n",
			   bytes_processed, num_docs,
			   ElapsedTime (&StartTime, NULL));
		exit (1);
	      }

#ifdef HAVE_TIMES
	    times (&tims);
	    pd->process_time += tims.tms_utime + tims.tms_stime;
#else
	    getrusage (RUSAGE_SELF, &ru);
	    tv->tv_sec += ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	    tv->tv_usec += ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
#endif
	  }
      num_docs++;
      if (Trace)
	{
	  block_bytes += (look_pos - base);
	  if (block_bytes >= trace)
	    {
#ifdef HAVE_MALLINFO
	      struct mallinfo mi;
	      mi = mallinfo ();
	      block_bytes -= trace;
	      fprintf (Trace, "%11" ULL_FS " bytes |%7lu docs |%7.3f Mb | %s\n",
		       bytes_processed, num_docs, mi.arena / 1024.0 / 1024.0,
		       ElapsedTime (&StartTime, NULL));
#else
	      block_bytes -= trace;
	      fprintf (Trace, "%11" ULL_FS " bytes |%7lu docs | %s\n",
		       bytes_processed, num_docs,
		       ElapsedTime (&StartTime, NULL));
#endif
	    }
	}
      if (!num && look_pos - 1 == end_pos)
	break;
    }

#ifndef HAVE_TIMES
  for (pass = 0; pass < MAX_PASSES; pass++)
    if (Passes & (1 << pass))
      time_normalise (&PassData[pass].process_time);
#endif

  GetTime (&ProcTime);

  for (pass = 0; pass < MAX_PASSES; pass++)
    if (Passes & (1 << pass))
      {
	pass_data *pd = &PassData[pass];
#ifdef HAVE_TIMES
	struct tms tims;
	times (&tims);
	pd->done_time -= tims.tms_utime + tims.tms_stime;
#else
	struct rusage ru;

	getrusage (RUSAGE_SELF, &ru);
	pd->done_time.tv_sec -= ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	pd->done_time.tv_usec -= ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
#endif
	if (pd->done (file_name) == COMPERROR)
	  FatalError (1, "Error during done of \"%s\"", pd->name);

#ifdef HAVE_TIMES
	times (&tims);
	pd->done_time += tims.tms_utime + tims.tms_stime;
#else
	getrusage (RUSAGE_SELF, &ru);
	pd->done_time.tv_sec += ru.ru_utime.tv_sec + ru.ru_stime.tv_sec;
	pd->done_time.tv_usec += ru.ru_utime.tv_usec + ru.ru_stime.tv_usec;
	time_normalise (&pd->done_time);
#endif
      }
  if (Trace)
    {
#ifdef HAVE_MALLINFO
      struct mallinfo mi;
      mi = mallinfo ();
      fprintf (Trace, "%11" ULL_FS " bytes |%7lu docs |%7.3f Mb | %s\n",
	       bytes_processed, num_docs, mi.arena / 1024.0 / 1024.0,
	       ElapsedTime (&StartTime, NULL));
#else
      fprintf (Trace, "%11" ULL_FS " bytes |%7lu docs | %s\n",
	       bytes_processed, num_docs,
	       ElapsedTime (&StartTime, NULL));
#endif
    }

  GetTime (&DoneTime);

  Message ("");
  Message ("%10s      :    init         process        done", "");
  for (pass = 0; pass < MAX_PASSES; pass++)
    if (Passes & (1 << pass))
      {
	pass_data *pd = &PassData[pass];
	char it[15], pt[15], dt[15];
#ifdef HAVE_TIMES
	strcpy (it, cputime_string (pd->init_time));
	strcpy (pt, cputime_string (pd->process_time));
	strcpy (dt, cputime_string (pd->done_time));
#else
	strcpy (it, cputime_string (&pd->init_time));
	strcpy (pt, cputime_string (&pd->process_time));
	strcpy (dt, cputime_string (&pd->done_time));
#endif
	Message ("%-10s      : %s   %s   %s", pd->name, it, pt, dt);
      }
  Message ("");
  Message ("Init time       : %s", ElapsedTime (&StartTime, &InitTime));
  Message ("Process time    : %s", ElapsedTime (&InitTime, &ProcTime));
  Message ("Done time       : %s", ElapsedTime (&ProcTime, &DoneTime));
  Message ("Total time      : %s", ElapsedTime (&StartTime, &DoneTime));
  Message ("Documents       : %u", num_docs);
  Message ("Bytes received  : %" ULL_FS, bytes_received);
  Message ("Bytes processed : %" ULL_FS, bytes_processed);
  Message ("Process Rate    : %.1f kB per cpu second",
   (double) bytes_processed / (ProcTime.CPUTime - InitTime.CPUTime) / 1024);
  free (buffer);
}



int 
main (int argc, char **argv)
{
  int ch, in_fd;
  char *filename = NULL;
  FILE *Trace = NULL;

  msg_prefix = argv[0];

  opterr = 0;
  while ((ch = getopt (argc, argv, "hC:WGSD123f:d:b:T:I:t:m:N:c:n:s:")) != -1)
    {
      switch (ch)
	{
	case 'G':
	  SkipSGML = 1;
	  break;
	case 'S':
	  Passes |= SPECIAL;
	  break;
	case '1':
	  InvfLevel = 1;
	  break;
	case '2':
	  InvfLevel = 2;
	  break;
	case '3':
	  InvfLevel = 3;
	  break;
	case 'f':
	  filename = optarg;
	  break;
	case 'n':
	  trace_name = optarg;
	  break;
	case 'D':
	  Dump = 1;
	  break;
	case 'W':
	  MakeWeights = 1;
	  break;
	case 'd':
	  set_basepath (optarg);
	  break;
	case 's':
	  stem_method = atoi (optarg) & STEMMER_MASK;
	  break;
	case 'b':
	  buf_size = atoi (optarg) * 1024;
	  break;
	case 'C':
	  comp_stat_point = atoi (optarg) * 1024;
	  break;
	case 'c':
	  ChunkLimit = atoi (optarg);
	  break;
	case 'm':
	  invf_buffer_size = (int) (atof (optarg) * 1024 * 1024);
	  break;
	case 'I':
	case 'N': /* N kept for compatability */
	  if (*optarg == '1')
	    Passes |= IVF_PASS_1;
	  else if (*optarg == '2')
	    Passes |= IVF_PASS_2;
	  else
	    usage ("Invalid pass number");
	  break;
	case 'T':
	  if (*optarg == '1')
	    Passes |= TEXT_PASS_1;
	  else if (*optarg == '2')
	    Passes |= TEXT_PASS_2;
	  else
	    usage ("Invalid pass number");
	  break;
	case 't':
	  trace = (unsigned long) (atof (optarg) * 1024 * 1024);
	  break;
	case 'h':
	case '?':
	  usage (NULL);
	}
    }

  if (!filename || *filename == '\0')
    FatalError (1, "A document collection name must be specified.");

  if (buf_size < MIN_BUF)
    FatalError (1, "The buffer size must exceed 1024 bytes.");

  if ((Passes & (IVF_PASS_1 | IVF_PASS_2)) == (IVF_PASS_1 | IVF_PASS_2))
    FatalError (1, "I1 and I2 cannot be done simultaneously.");

  if ((Passes & (TEXT_PASS_1 | TEXT_PASS_2)) == (TEXT_PASS_1 | TEXT_PASS_2))
    FatalError (1, "T1 and T2 cannot be done simultaneously.");

  if (!Passes)
    FatalError (1, "S, T1, T2, I1 or I2 must be specified.");

  if (optind < argc)
    {
      if ((in_fd = open (argv[optind], O_RDONLY)) == -1)
	FatalError (1, "Cannot open %s", argv[optind]);
      files = &argv[optind + 1];
      num_files = argc - (optind + 1);
    }
  else
    in_fd = 0;			/* stdin */


  if (trace)
    {
      if (!trace_name)
	trace_name = make_name (filename, TRACE_SUFFIX, NULL);
      if (!(Trace = fopen (trace_name, "a")))
	Message ("Unable to open \"%s\". No tracing will be done.", trace_name);
      else
	setbuf (Trace, NULL);
    }
  else
    Trace = NULL;

  if (comp_stat_point)
    {
      char *name = make_name (filename, COMPRESSION_STATS_SUFFIX, NULL);
      if (!(Comp_Stats = fopen (name, "w")))
	Message ("Unable to open \"%s\". No comp. stats. will be generated.",
		 name);
    }


  if (Trace)
    {
      int i;
      fprintf (Trace, "\n\n\t\t-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-\n\n");
      for (i = 0; i < argc; i++)
	fprintf (Trace, "%s ", argv[i]);
      fprintf (Trace, "\n\n");
    }

  driver (in_fd, Trace, filename);

  if (Trace)
    fclose (Trace);

  if (Comp_Stats)
    fclose (Comp_Stats);

  exit (0);
}
